File: chapter 07/recipe 7.5/DniNetCurrencyVB/DniNetCurrencyObj.vb
Imports System.Runtime.InteropServices

Public Interface ICurrency
    Function UseNativeVariant(ByVal inParam As Object, _
        ByRef outParam As Object) As String
    Function UseVariantCurrency(ByVal inParam As Object, _
        ByRef outParam As Object) As String
    Function UseDecimalCurrency( _
        <MarshalAs(UnmanagedType.Currency)> _
        ByVal inParam As Decimal, _
        <MarshalAs(UnmanagedType.Currency)> _
        ByRef outParam As Decimal) As String
End Interface

<ClassInterface(ClassInterfaceType.None)> _
Public Class DniNetCurrencyObj
    Implements ICurrency

    Public Function UseNativeVariant(ByVal inParam As Object, _
        ByRef outParam As Object) As String _
            Implements ICurrency.UseNativeVariant
        outParam = inParam
        Return inParam.GetType().Name
    End Function

    Public Function UseVariantCurrency(ByVal inParam As Object, _
        ByRef outParam As Object) As String _
            Implements ICurrency.UseVariantCurrency
        outParam = New CurrencyWrapper(inParam)
        Return inParam.GetType().Name
    End Function

    Public Function UseDecimalCurrency(ByVal inParam As Decimal, _
        ByRef outParam As Decimal) As String _
            Implements ICurrency.UseDecimalCurrency
        outParam = inParam
        Return inParam.GetType().Name
    End Function

End Class

File: chapter 07/recipe 7.5/DniNetCurrency/DniNetCurrencyObj.cs
using System;
using System.Runtime.InteropServices;

namespace DniNetCurrency
{
	public interface ICurrency
	{
		string UseNativeVariant(Object inParam, ref Object outParam);

		string UseVariantCurrency(Object inParam, ref Object outParam);

		string UseDecimalCurrency(
			[MarshalAs(UnmanagedType.Currency)]
			Decimal inParam, 
			[MarshalAs(UnmanagedType.Currency)]
			ref Decimal outParam);
	}

	[ClassInterface(ClassInterfaceType.None)]
	public class DniNetCurrencyObj : ICurrency
	{
		public string UseNativeVariant(
			Object inParam, ref Object outParam)
		{
			outParam = inParam;
			return inParam.GetType().Name;
		}

		public string UseVariantCurrency(
			Object inParam, ref Object outParam)
		{
			outParam = new CurrencyWrapper(inParam);
			return inParam.GetType().Name;
		}

		public string UseDecimalCurrency(
			Decimal inParam, ref Decimal outParam)
		{
			outParam = inParam;
			return inParam.GetType().Name;
		}
	}
}
